pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,179 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	Parent   string
+	depth    int
+	nodeType string // packageName, packageVersion
+}
+
+type queueValues struct {
+	path    []string
+	nodeMap map[string]DfsNode
+	now     string
+	nowNode DfsNode
+	queue   []string
+}
+
+// TODO: make more robust using predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, startType string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%v", err)
+	}
+
+	_, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+
+	q := queueValues{
+		queue:   make([]string, 0), // the queue of nodes in bfs
+		nodeMap: map[string]DfsNode{},
+	}
+
+	q.nodeMap[startID] = DfsNode{
+		nodeType: startType,
+	}
+
+	q.queue = append(q.queue, startID)
+
+	for len(q.queue) > 0 {
+		q.now = q.queue[0]
+		q.queue = q.queue[1:]
+		q.nowNode = q.nodeMap[q.now]
+
+		if stopID == q.now {
+			break
+		}
+
+		if q.nowNode.depth >= maxDepth {
+			break
+		}
+
+		neighborsResponse, err := model.Neighbors(ctx, gqlclient, q.now, []model.Edge{})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed getting neighbors:%w", err)
+		}
+
+		for _, neighbor := range neighborsResponse.Neighbors {
+			q, err = caseOnPredicates(ctx, gqlclient, &q, neighbor)
+
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		q.nowNode.expanded = true
+		q.nodeMap[q.now] = q.nowNode
+	}
+
+	return q.nodeMap, nil
+
+}
+
+func caseOnPredicates(ctx context.Context, gqlclient graphql.Client, q *queueValues, neighbor model.NeighborsNeighborsNode) (queueValues, error) {
+	// case on predicates and nodeType
+	switch q.nowNode.nodeType {
+	case "packageVersion":
```

Comment:
technically this shouldnt find any `IsDeps` nodes since it attaches to `packageName` for dependents

Reply:
Resolved, changed to packageName.